feat: add bundle `--component-names-strategy` flag (#2851)
Redocly/redocly-cli: docs/@v2/commands/bundle.md

| --remove-unused-components              | boolean  | Remove unused components from the `bundle` output.                                                                                                                                                                                                        |
| --skip-decorator                        | [string] | Ignore certain decorators. See the [Skip preprocessor, rule, or decorator section](#skip-preprocessor-rule-or-decorator).                                                                                                                                 |
| --skip-preprocessor                     | [string] | Ignore certain preprocessors. See the [Skip preprocessor, rule, or decorator section](#skip-preprocessor-rule-or-decorator).                                                                                                                              |

> | --use-titles-for-component-names        | boolean  | Build Schema component names from each schema's `title` field. See [Use titles for component names](#use-titles-for-component-names).                                                                                                                     |

Reviewer: Let's change it to --component-names-strategy with options like "basename" (default) or "title"?

@tatomyr what do you think?

Reviewer 2: Sounds even better to me. What about "filename" instead of "basename"?

Reviewer: I'm fine with either

- `off`: No warnings or errors are shown.
- `warn` (default): Shows a warning and renames conflicting components automatically.
- `error`: Treats conflicts as errors; the bundling process fails if a naming conflict is detected.

### Use titles for component names

> By default, the bundler names each inlined Schema component after its `$ref` — the JSON Pointer fragment, or the file's basename when there's no fragment. Two files that share a basename — for example `schemas/models/Authority.yaml` and `schemas/requests/Authority.yaml` — collide into `Authority` and `Authority-2`. That auto-numbered suffix is brittle: an unrelated `$ref` change can renumber it.

Reviewer 3: ```suggestion
By default, the bundler names each inlined Schema component after its `$ref` — the JSON Pointer fragment, or the file's basename when there's no fragment.
Two files that share a basename — for example `schemas/models/Authority.yaml` and `schemas/requests/Authority.yaml` — collide into `Authority` and `Authority-2`.
The auto-numbered suffix is brittle: an unrelated `$ref` change can renumber it.
```

With `--use-titles-for-component-names`, the name comes from each schema's `title` instead. Words (split on spaces) are capitalized and joined, and the spec-legal `.`, `-`, and `_` are kept:

```yaml
# schemas/models/Authority.yaml
type: object
title: Authority model

# schemas/requests/Authority.yaml
type: object
title: Authority request
```

```bash
redocly bundle openapi.yaml -o bundled.yaml --use-titles-for-component-names
```

> The output now uses `AuthorityModel` and `AuthorityRequest`. A `title` containing `.`, `-`, or `_` keeps them — `order-item` becomes `Order-item`.

Reviewer 3: ```suggestion
The output uses `AuthorityModel` and `AuthorityRequest`.
A `title` containing `.`, `-`, or `_` keeps them — `order-item` becomes `Order-item`.
```

### Use titles for component names

By default, the bundler names each inlined Schema component after its `$ref` — the JSON Pointer fragment, or the file's basename when there's no fragment.
Two files that share a basename — for example `schemas/models/Authority.yaml` and `schemas/requests/Authority.yaml` — collide into `Authority` and `Authority-2`.
The auto-numbered suffix is brittle: an unrelated `$ref` change can renumber it.

With `--use-titles-for-component-names`, the name comes from each schema's `title` instead. Words (split on spaces) are capitalized and joined, and the spec-legal `.`, `-`, and `_` are kept:

```yaml
# schemas/models/Authority.yaml
type: object
title: Authority model

# schemas/requests/Authority.yaml
type: object
title: Authority request
```

```bash
redocly bundle openapi.yaml -o bundled.yaml --use-titles-for-component-names
```

The output uses `AuthorityModel` and `AuthorityRequest`.
A `title` containing `.`, `-`, or `_` keeps them — `order-item` becomes `Order-item`.

Bundling fails (no output file unless you pass `--force`) when:

- A referenced schema has no `title`.

> - A `title` uses anything other than ASCII letters, digits, spaces, `.`, `-`, or `_` (for example `&` or a non-ASCII title).

Reviewer 2: Could you explain your consideration for choosing this characters set?

Author: The character it's the Components Object key pattern from the `OpenAPI` and `AsyncAPI` specs `(^[a-zA-Z0-9.\-_]+$)`, the same one the `spec-components-invalid-map-name` rule uses.

```yaml
# schemas/models/Order.yaml
type: object
title: Order model

# schemas/requests/Order.yaml
type: object
title: Order request
```

{% tabs %}
{% tab label="basename (default)" %}

Names come from the `$ref` — the JSON Pointer fragment, or the file's basename when there's no fragment.
Because both files are `Order.yaml`, they collide and the duplicate is auto-numbered:

```bash
redocly bundle openapi.yaml -o bundled.yaml --component-names-strategy=basename
```

The output uses `Order` and `Order-2`.
The auto-numbered suffix is brittle: an unrelated `$ref` change can renumber it.

{% /tab  %}
{% tab label="title" %}

Names come from each schema's `title`, converted to PascalCase (words are split on spaces, capitalized, and joined), so they don't depend on file paths or `$ref` order:

```bash
redocly bundle openapi.yaml -o bundled.yaml --component-names-strategy=title
```

The output uses `OrderModel` and `OrderRequest`.

{% /tab  %}
{% /tabs  %}

{% admonition type="warning" name="Title sanitization" %}
The OpenAPI and AsyncAPI specifications allow only ASCII letters (`a`–`z`, `A`–`Z`), digits, `.`, `-`, and `_` in a Components Object key.

> Those characters are kept; any other character — including non-ASCII letters such as `é` or `я` — is replaced with `-` (for example, `User & Group` becomes `User-Group`).

Reviewer 3: ```suggestion
All other characters, including non-ASCII letters such as `é` or `я`, are replaced with `-` (for example, `User & Group` becomes `User-Group`).
```

> A schema without a `title` can't be named this way, so bundling reports an error for that schema.

Reviewer 3: ```suggestion
Schemas without `title` can't be named using the `--component-names-strategy=title` strategy.
The bundling process reports an error for such schemas.
```
